File: iOS-SDK/PhoneAgentDemo/SDK/Agent/AgentEngine.swift
//
//  AgentEngine.swift
//  PhoneAgentSDK
//
//  Created by AutoGLM on 2025-01-30.
//

import UIKit
import OSLog

/// AI Agent 引擎 - 核心协调器
let agentLogger = Logger(subsystem: "com.demo.PhoneAgentDemo", category: "AgentEngine")
@MainActor
public class AgentEngine: ObservableObject {

    // MARK: - Published Properties

    /// 当前步骤
    @Published public private(set) var currentStep = 0

    /// 是否正在执行
    @Published public private(set) var isRunning = false

    /// 当前思考内容
    @Published public private(set) var currentThinking = ""

    /// 当前动作
    @Published public private(set) var currentAction = ""

    /// 执行历史
    @Published public private(set) var history: [AgentResult.StepInfo] = []

    /// 最后的错误信息
    @Published public private(set) var lastError: Error?

    // MARK: - Properties

    private let modelClient: ModelClient
    private let actionExecutor: ActionExecutor
    private let config: AgentConfig

    private var messages: [MessageBuilder.Message] = []
    private var conversationStartTime: Date?
    private var unknownActionCount = 0  // 连续未知动作的计数

    // MARK: - Callbacks

    /// 步骤进度回调
    public var onStepProgress: ((Int, String, String) -> Void)?

    /// 敏感操作确认回调
    public var onSensitiveAction: ((String) async -> Bool)?

    /// 人工接管回调
    public var onTakeOver: ((String) -> Void)?

    /// 完成回调
    public var onComplete: ((AgentResult) -> Void)?

    // MARK: - Initialization

    public init(
        modelConfig: ModelConfig = .default,
        agentConfig: AgentConfig = .default
    ) {
        self.config = agentConfig
        self.modelClient = ModelClient(config: modelConfig)
        self.actionExecutor = ActionExecutor()

        setupCallbacks()
    }

    public init(
        modelClient: ModelClient,
        actionExecutor: ActionExecutor,
        config: AgentConfig = .default
    ) {
        self.modelClient = modelClient
        self.actionExecutor = actionExecutor
        self.config = config

        setupCallbacks()
    }

    private func setupCallbacks() {
        // 模型思考回调
        modelClient.onThinking = { [weak self] thinking in
            self?.currentThinking = thinking
            self?.log("思考: \(thinking)")
        }

        // 模型错误回调
        modelClient.onError = { [weak self] error in
            self?.lastError = error
            self?.log("模型错误: \(error.localizedDescription)")
        }

        // 敏感操作回调
        actionExecutor.setSensitiveActionCallback { [weak self] message in
            if let callback = self?.onSensitiveAction {
                return await callback(message)
            }
            // 默认自动确认
            return true
        }

        // 监听人工接管通知
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleTakeOver(_:)),
            name: .agentTakeOver,
            object: nil
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    // MARK: - Public Methods

    /// 运行任务
    /// - Parameter instruction: 用户指令
    /// - Returns: 执行结果
    public func run(instruction: String) async -> AgentResult {
        reset()
        isRunning = true
        conversationStartTime = Date()

        log("=== 开始执行任务 ===")
        log("指令: \(instruction)")

        do {
            // 首次步骤 - 添加用户指令
            try await executeFirstStep(instruction: instruction)

            // 循环执行直到完成或达到最大步数
            while currentStep < config.maxSteps {
                // 检查是否应该结束
                if let lastStep = history.last, lastStep.actionType.isFinished {
                    complete(success: true, message: lastStep.message)
                    return AgentResult.success(
                        message: lastStep.message ?? "任务完成",
                        totalSteps: currentStep,
                        history: history
                    )
                }

                // 执行下一步
                let shouldContinue = try await executeNextStep()
                if !shouldContinue {
                    break
                }
            }

            // 达到最大步数
            complete(success: false, message: "达到最大执行步数")
            return AgentResult.maxStepsReached(totalSteps: currentStep, history: history)

        } catch {
            log("执行失败: \(error.localizedDescription)")
            complete(success: false, message: error.localizedDescription)
            return AgentResult.failure(
                message: error.localizedDescription,
                totalSteps: currentStep,
                history: history
            )
        }
    }

    /// 单步执行（用于手动控制）
    /// - Parameter instruction: 用户指令（首次步骤需要）
    /// - Returns: 执行结果
    public func step(instruction: String? = nil) async throws -> AgentResult.StepInfo? {
        if currentStep == 0 {
            guard let instruction = instruction else {
                throw ModelError.parsingError
            }
            try await executeFirstStep(instruction: instruction)
        } else {
            _ = try await executeNextStep()
        }

        return history.last
    }

    /// 停止执行
    public func stop() {
        isRunning = false
        modelClient.cancel()
        log("=== 任务已停止 ===")
    }

    /// 重置状态
    public func reset() {
        currentStep = 0
        isRunning = false
        currentThinking = ""
        currentAction = ""
        history = []
        lastError = nil
        messages = []
        conversationStartTime = nil
        unknownActionCount = 0
    }

    // MARK: - Private Methods - 执行步骤

    private func executeFirstStep(instruction: String) async throws {
        // 添加系统提示词
        let systemMessage = MessageBuilder.createSystemMessage(
            content: MessageBuilder.getSystemPrompt(
                language: modelClient.configValue.language
            )
        )
        messages.append(systemMessage)

        // 捕获屏幕
        let screenshot = try ScreenCapture.capture()
        let screenName = ScreenCapture.getCurrentScreenName()

        // 构建屏幕信息
        let screenInfo = ScreenInfo(screenshot: screenshot, screenName: screenName)
        let screenText = MessageBuilder.buildScreenInfo(
            currentScreen: screenName,
            screenWidth: screenshot.width,
            screenHeight: screenshot.height
        )

        // 添加用户消息
        let userText = "\(instruction)\n\n** 屏幕信息 **\n\n\(screenText)"
        let userMessage = MessageBuilder.createUserMessage(
            text: userText,
            imageBase64: screenshot.base64Data
        )
        messages.append(userMessage)

        // 调用模型
        try await callModelAndExecute(
            screenshot: screenshot,
            isFirstStep: true
        )
    }

    private func executeNextStep() async throws -> Bool {
        currentStep += 1

        // 检查上一步是否完成
        if let lastStep = history.last, lastStep.actionType.isFinished {
            return false
        }

        // 短暂延迟
        try await Task.sleep(nanoseconds: UInt64(config.stepDelay * 1_000_000_000))

        // 捕获屏幕
        let screenshot = try ScreenCapture.capture()
        let screenName = ScreenCapture.getCurrentScreenName()

        // 构建屏幕信息
        let screenText = MessageBuilder.buildScreenInfo(
            currentScreen: screenName,
            screenWidth: screenshot.width,
            screenHeight: screenshot.height
        )

        // 添加用户消息
        let userText = "** 屏幕信息 **\n\n\(screenText)"
        let userMessage = MessageBuilder.createUserMessage(
            text: userText,
            imageBase64: screenshot.base64Data
        )
        messages.append(userMessage)

        // 调用模型
        try await callModelAndExecute(
            screenshot: screenshot,
            isFirstStep: false
        )

        return true
    }

    private func callModelAndExecute(
        screenshot: ScreenCapture.Screenshot,
        isFirstStep: Bool
    ) async throws {
        let startTime = Date()

        // 调用模型
        let response = try await modelClient.request(
            messages: messages,
            screenWidth: screenshot.width,
            screenHeight: screenshot.height
        )

        // 检查响应是否为空
        if response.action.isEmpty {
            log("⚠️ 模型返回空动作，停止执行")
            throw ModelError.parsingError
        }

        // 解析动作
        let actionType = response.parseAction(
            screenWidth: screenshot.width,
            screenHeight: screenshot.height
        )

        currentAction = response.action

        // 检查是否为未知动作
        if case .unknown(let text) = actionType {
            unknownActionCount += 1
            log("⚠️ 未知动作 (\(unknownActionCount)/3): \(text.prefix(100))")

            // 连续 3 次未知动作时终止
            if unknownActionCount >= 3 {
                log("❌ 连续 3 次解析失败，终止执行")
                lastError = ModelError.parsingError
                throw ModelError.parsingError
            }
        } else {
            // 成功解析，重置计数
            if unknownActionCount > 0 {
                unknownActionCount = 0
            }
        }

        // 移除上下文中的图片以节省空间
        messages[messages.count - 1] = MessageBuilder.removeImages(from: messages[messages.count - 1])

        // 添加助手响应到上下文
        let assistantMessage = MessageBuilder.createAssistantMessage(
            content: "<thinking>\(response.thinking)</thinking><answer>\(response.action)</answer>"
        )
        messages.append(assistantMessage)

        // 执行动作
        let executionResult = try await actionExecutor.execute(actionType)

        let duration = Date().timeIntervalSince(startTime)

        // 记录步骤
        let stepInfo = AgentResult.StepInfo(
            step: currentStep,
            thinking: response.thinking,
            action: response.action,
            actionType: actionType,
            success: executionResult.success,
            message: executionResult.message,
            duration: duration
        )
        history.append(stepInfo)

        // 通知进度
        onStepProgress?(currentStep, response.thinking, response.action)

        // 执行失败时记录
        if !executionResult.success {
            log("⚠️ 动作执行失败: \(executionResult.message ?? "未知错误")")
        }
    }

    // MARK: - Private Methods - 辅助

    private func complete(success: Bool, message: String?) {
        isRunning = false
        log("=== 任务\(success ? "完成" : "失败") ===")
        if let message = message {
            log("结果: \(message)")
        }

        let result = AgentResult(
            success: success,
            message: message,
            totalSteps: currentStep,
            history: history,
            metrics: calculateMetrics()
        )

        onComplete?(result)
    }

    private func calculateMetrics() -> AgentResult.ExecutionMetrics? {
        guard let startTime = conversationStartTime else { return nil }

        let totalDuration = Date().timeIntervalSince(startTime)
        let inferenceTime = history.reduce(0) { $0 + $1.duration }
        let executionTime = totalDuration - inferenceTime

        return AgentResult.ExecutionMetrics(
            totalDuration: totalDuration,
            averageStepDuration: history.isEmpty ? 0 : totalDuration / Double(history.count),
            totalInferenceTime: inferenceTime,
            totalExecutionTime: executionTime
        )
    }

    private func log(_ message: String) {
        if config.verbose {
            agentLogger.error("\(message)")
        }
    }

    // MARK: - Notification Handlers

    @objc private func handleTakeOver(_ notification: Notification) {
        guard let message = notification.userInfo?["message"] as? String else {
            return
        }
        onTakeOver?(message)
    }
}

// MARK: - ModelResponse.ActionType 扩展
extension ModelResponse.ActionType {
    var isFinished: Bool {
        if case .finish = self {
            return true
        }
        return false
    }
}

File: Android-SDK/phoneagent-sdk/src/main/java/com/autoglm/phoneagent/agent/AgentEngine.kt
package com.autoglm.phoneagent.agent

import android.app.Activity
import android.app.Application
import android.os.Bundle
import com.autoglm.phoneagent.action.ActionExecutor
import com.autoglm.phoneagent.action.ActionType
import com.autoglm.phoneagent.action.ExecutionResult
import com.autoglm.phoneagent.capture.ScreenCapture
import com.autoglm.phoneagent.capture.ScreenInfo
import com.autoglm.phoneagent.model.MessageBuilder
import com.autoglm.phoneagent.model.ModelClient
import com.autoglm.phoneagent.model.ModelConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

/**
 * Activity lifecycle callbacks to track current foreground activity.
 */
private class ActivityTracker : Application.ActivityLifecycleCallbacks {
    private var currentActivityRef: WeakReference<Activity>? = null

    fun getCurrentActivity(): Activity? = currentActivityRef?.get()

    fun setCurrentActivity(activity: Activity) {
        currentActivityRef = WeakReference(activity)
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}

    override fun onActivityStarted(activity: Activity) {
        setCurrentActivity(activity)
    }

    override fun onActivityResumed(activity: Activity) {
        setCurrentActivity(activity)
    }

    override fun onActivityPaused(activity: Activity) {}

    override fun onActivityStopped(activity: Activity) {}

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}

    override fun onActivityDestroyed(activity: Activity) {}
}

/**
 * Main engine for the Phone Agent SDK.
 * Coordinates screenshot capture, model communication, and action execution.
 *
 * @property activity The activity to operate on
 * @property modelClient Client for communicating with the model API
 * @property config Agent configuration
 */
class AgentEngine(
    private val activity: Activity,
    private val modelClient: ModelClient,
    private val config: AgentConfig = AgentConfig.DEFAULT
) {
    // Pass getCurrentActivity as a function reference to ActionExecutor
    // This allows the executor to dynamically get the current activity
    private val actionExecutor = ActionExecutor { getCurrentActivity() }

    // Track current foreground activity for correct screenshot capture
    private val activityTracker = ActivityTracker()

    init {
        // Set initial activity
        activityTracker.setCurrentActivity(activity)
        // Register lifecycle callbacks to track activity changes
        (activity.applicationContext as? Application)?.registerActivityLifecycleCallbacks(activityTracker)
    }

    /**
     * Get the current foreground activity.
     * Falls back to the original activity if tracking fails.
     */
    private fun getCurrentActivity(): Activity {
        return activityTracker.getCurrentActivity() ?: activity
    }

    /**
     * Clean up resources when engine is stopped.
     */
    fun cleanup() {
        (activity.applicationContext as? Application)?.unregisterActivityLifecycleCallbacks(activityTracker)
    }

    // State
    private val _currentStep = MutableStateFlow(0)
    private val _isRunning = MutableStateFlow(false)
    private val _currentThinking = MutableStateFlow("")
    private val _currentAction = MutableStateFlow("")
    private val _history = MutableStateFlow<List<AgentResult.StepInfo>>(emptyList())
    private val _error = MutableStateFlow<String?>(null)

    private var executionJob: Job? = null
    private var messages: MutableList<MessageBuilder.Message> = mutableListOf()
    private var unknownActionCount = 0

    // Helper method for conditional logging
    private fun logDebug(message: String) {
        if (config.verbose) {
            println("[AgentEngine] $message")
        }
    }

    private fun logInfo(message: String) {
        println("[AgentEngine] $message")
    }

    private fun logError(message: String) {
        println("[AgentEngine] ERROR: $message")
    }

    // Public state flows
    val currentStep: StateFlow<Int> = _currentStep.asStateFlow()
    val isRunning: StateFlow<Boolean> = _isRunning.asStateFlow()
    val currentThinking: StateFlow<String> = _currentThinking.asStateFlow()
    val currentAction: StateFlow<String> = _currentAction.asStateFlow()
    val history: StateFlow<List<AgentResult.StepInfo>> = _history.asStateFlow()
    val error: StateFlow<String?> = _error.asStateFlow()

    // Callbacks
    var onStepProgress: ((step: Int, thinking: String, action: String) -> Unit)? = null
    var onSensitiveAction: ((message: String) -> suspend () -> Boolean)? = null
    var onTakeOver: ((message: String) -> Unit)? = null
    var onComplete: ((AgentResult) -> Unit)? = null

    /**
     * Run the agent with the given instruction.
     *
     * @param instruction The user's task instruction
     * @return AgentResult with the execution outcome
     */
    suspend fun run(instruction: String): AgentResult = withContext(Dispatchers.Main) {
        logInfo("开始执行任务: $instruction")

        if (_isRunning.value) {
            logError("Agent 已在运行中")
            return@withContext AgentResult.failure("Agent 已在运行中")
        }

        config.validate()

        // Reset state
        resetState()
        _isRunning.value = true

        val startTime = System.currentTimeMillis()
        val historyList = mutableListOf<AgentResult.StepInfo>()
        var totalInferenceTime = 0L
        var totalExecutionTime = 0L

        try {
            // Execute first step with instruction
            val firstStepResult = executeFirstStep(instruction)
            if (firstStepResult != null) {
                historyList.add(firstStepResult)
                totalInferenceTime += firstStepResult.inferenceTime
                totalExecutionTime += firstStepResult.executionTime
            }

            // Continue executing steps
            while (_currentStep.value < config.maxSteps && isActive) {
                // Check if previous step finished the task
                val lastStep = historyList.lastOrNull()
                if (lastStep?.actionType == ActionType.FINISH) {
                    logInfo("任务完成")
                    break
                }

                // Check if user takeover was requested
                if (lastStep?.actionType == ActionType.TAKE_OVER) {
                    onTakeOver?.invoke(lastStep.message ?: "需要用户协助")
                    break
                }

                // Execute next step
                val stepResult = executeNextStep()
                if (stepResult != null) {
                    historyList.add(stepResult)
                    totalInferenceTime += stepResult.inferenceTime
                    totalExecutionTime += stepResult.executionTime
                } else {
                    break
                }
            }

            val totalTime = System.currentTimeMillis() - startTime
            val lastStep = historyList.lastOrNull()
            val finalHistory = historyList.toList()

            // Build metrics
            val metrics = if (historyList.isNotEmpty()) {
                AgentResult.ExecutionMetrics(
                    totalDuration = totalTime,
                    averageStepDuration = if (historyList.isNotEmpty()) {
                        totalTime.toDouble() / historyList.size
                    } else 0.0,
                    totalInferenceTime = totalInferenceTime,
                    totalExecutionTime = totalExecutionTime
                )
            } else null

            // Determine final result
            val result = when {
                lastStep?.actionType == ActionType.FINISH -> AgentResult.success(
                    message = lastStep.message,
                    totalSteps = historyList.size,
                    history = finalHistory,
                    metrics = metrics
                )

                lastStep?.actionType == ActionType.TAKE_OVER -> AgentResult.userTakeover(
                    message = lastStep.message ?: "用户接管了任务",
                    totalSteps = historyList.size,
                    history = finalHistory,
                    metrics = metrics
                )

                _currentStep.value >= config.maxSteps -> AgentResult.maxStepsReached(
                    totalSteps = historyList.size,
                    history = finalHistory,
                    metrics = metrics
                )

                else -> AgentResult.failure(
                    message = "任务未完成",
                    totalSteps = historyList.size,
                    history = finalHistory,
                    metrics = metrics
                )
            }

            onComplete?.invoke(result)
            result

        } catch (e: Exception) {
            val errorResult = AgentResult.failure(
                message = "执行错误: ${e.message}",
                totalSteps = historyList.size,
                history = historyList.toList()
            )
            _error.value = e.message
            onComplete?.invoke(errorResult)
            errorResult
        } finally {
            _isRunning.value = false
        }
    }

    /**
     * Execute the first step with the instruction.
     */
    private suspend fun executeFirstStep(instruction: String): AgentResult.StepInfo? {
        // Build messages with system prompt and instruction
        messages.clear()
        messages.add(MessageBuilder.buildSystemMessage(modelClient.config.language))

        // Capture screenshot
        val currentActivity = getCurrentActivity()
        val screenshot = ScreenCapture.captureActivity(currentActivity)
        val screenInfo = ScreenInfo.fromActivity(currentActivity, includeElements = false)
        logDebug("截图: ${screenshot.width}x${screenshot.height}, Activity: ${currentActivity.javaClass.simpleName}")

        // Add user message with instruction and screenshot
        messages.add(
            MessageBuilder.buildUserMessage(
                instruction = instruction,
                screenshotBase64 = screenshot.toDataURL(),
                screenInfo = screenInfo
            )
        )

        return callModelAndExecute()
    }

    /**
     * Execute subsequent steps.
     */
    private suspend fun executeNextStep(): AgentResult.StepInfo? {
        // Wait for UI to update
        delay(config.stepDelay)

        // Capture new screenshot
        val currentActivity = getCurrentActivity()
        val screenshot = ScreenCapture.captureActivity(currentActivity)
        val screenInfo = ScreenInfo.fromActivity(currentActivity, includeElements = false)

        // Add follow-up message with screenshot only
        messages.add(
            MessageBuilder.buildFollowUpMessage(
                screenshotBase64 = screenshot.toDataURL(),
                screenInfo = screenInfo
            )
        )

        return callModelAndExecute()
    }

    /**
     * Call the model and execute the returned action.
     */
    private suspend fun callModelAndExecute(): AgentResult.StepInfo? {
        val stepStartTime = System.currentTimeMillis()

        val response = try {
            modelClient.sendRequest(messages.toList())
        } catch (e: Exception) {
            logError("模型请求异常: ${e.javaClass.simpleName}: ${e.message}")
            e.printStackTrace()
            _error.value = "模型请求失败: ${e.message}"
            return AgentResult.StepInfo(
                step = _currentStep.value + 1,
                thinking = "错误",
                action = "无",
                actionType = ActionType.UNKNOWN,
                success = false,
                message = e.message,
                duration = System.currentTimeMillis() - stepStartTime
            )
        }

        logDebug("步骤 ${_currentStep.value + 1}: ${response.parsedAction.type}")

        val inferenceTime = response.totalTime
        _currentThinking.value = response.thinking
        _currentAction.value = response.action

        // Check for unknown action
        if (response.parsedAction.type == ActionType.UNKNOWN) {
            unknownActionCount++
            logError("未知动作 #$unknownActionCount: ${response.parsedAction.rawAction}")

            if (unknownActionCount >= config.unknownActionThreshold) {
                return AgentResult.StepInfo(
                    step = _currentStep.value + 1,
                    thinking = response.thinking,
                    action = response.action,
                    actionType = ActionType.UNKNOWN,
                    success = false,
                    message = "连续 $unknownActionCount 次未知动作，终止执行",
                    duration = System.currentTimeMillis() - stepStartTime
                )
            }
        } else {
            unknownActionCount = 0
        }

        // Handle sensitive actions (tap with message)
        if (response.parsedAction.type == ActionType.TAP_WITH_MESSAGE) {
            val message = response.parsedAction.message ?: "敏感操作"
            val shouldProceed = if (config.autoConfirmSensitiveActions) {
                true
            } else {
                onSensitiveAction?.invoke(message)?.invoke() ?: true
            }

            if (!shouldProceed) {
                return AgentResult.StepInfo(
                    step = _currentStep.value + 1,
                    thinking = response.thinking,
                    action = response.action,
                    actionType = ActionType.TAKE_OVER,
                    success = false,
                    message = "用户取消了敏感操作",
                    duration = System.currentTimeMillis() - stepStartTime
                )
            }
        }

        // Execute the action
        val executionStart = System.currentTimeMillis()
        val executionResult = actionExecutor.execute(response.parsedAction)
        val executionTime = System.currentTimeMillis() - executionStart

        // Add extra delay for navigation and scroll actions to allow UI to update
        val navigationActions = setOf(
            ActionType.TAP,
            ActionType.TAP_WITH_MESSAGE,
            ActionType.DOUBLE_TAP,
            ActionType.BACK,
            ActionType.SWIPE
        )
        if (response.parsedAction.type in navigationActions && executionResult.isSuccess) {
            val delayTime = if (response.parsedAction.type == ActionType.SWIPE) 300 else 400
            delay(delayTime.toLong())
        }

        val stepNumber = _currentStep.value + 1
        _currentStep.value = stepNumber

        val stepInfo = AgentResult.StepInfo(
            step = stepNumber,
            thinking = response.thinking,
            action = response.action,
            actionType = response.parsedAction.type,
            success = executionResult.isSuccess || executionResult.shouldFinish,
            message = executionResult.getMessage(),
            duration = System.currentTimeMillis() - stepStartTime
        )

        // Update history
        val currentHistory = _history.value.toMutableList()
        currentHistory.add(stepInfo)
        _history.value = currentHistory

        // Notify callback
        onStepProgress?.invoke(stepNumber, response.thinking, response.action)

        // Log if verbose
        if (config.verbose) {
            val result = when {
                executionResult.shouldFinish -> "✓ (完成)"
                executionResult.isSuccess -> "✓"
                else -> "✗"
            }
            logDebug("步骤 $stepNumber: ${response.parsedAction.type} $result")
        }

        return stepInfo.copy(
            inferenceTime = inferenceTime,
            executionTime = executionTime
        )
    }

    /**
     * Reset the engine state.
     */
    private fun resetState() {
        _currentStep.value = 0
        _currentThinking.value = ""
        _currentAction.value = ""
        _history.value = emptyList()
        _error.value = null
        unknownActionCount = 0
        messages.clear()
    }

    /**
     * Stop the agent execution.
     */
    fun stop() {
        executionJob?.cancel()
        _isRunning.value = false
    }

    /**
     * Get the current execution history as a formatted string.
     */
    fun getHistorySummary(): String {
        return _history.value.joinToString("\n\n") { step ->
            """
            步骤 ${step.step}:
            思考: ${step.thinking}
            动作: ${step.action}
            结果: ${if (step.success) "✓ 成功" else "✗ 失败"}
            ${if (step.message != null) "消息: ${step.message}" else ""}
            """.trimIndent()
        }
    }
}

/**
 * Factory for creating AgentEngine instances.
 */
object AgentEngineFactory {
    /**
     * Create an AgentEngine with the given configuration.
     *
     * @param activity The activity to operate on
     * @param modelConfig Model API configuration
     * @param agentConfig Agent execution configuration
     * @return Configured AgentEngine instance
     */
    fun create(
        activity: Activity,
        modelConfig: ModelConfig = ModelConfig.DEFAULT,
        agentConfig: AgentConfig = AgentConfig.DEFAULT
    ): AgentEngine {
        modelConfig.validate()
        agentConfig.validate()

        val modelClient = ModelClient(modelConfig)
        return AgentEngine(activity, modelClient, agentConfig)
    }
}
